File: ViewModels/OnboardingViewModel.swift
import Foundation
import Combine
import UIKit
import FirebaseAnalytics

class OnboardingViewModel: ObservableObject {
    enum OnboardingStep: CaseIterable {
        case relationshipGoals
        case relationshipDate
        case relationshipImprovement
        case authentication
        case displayName
        case profilePhoto
        case completion
        case loading
        case partnerCode
        case fitnessIntro
        case fitnessIntro2
        case dailyQuestionNotification
        case categoriesPreview
        case subscription
    }
    
    @Published var currentStep: OnboardingStep = .relationshipGoals
    @Published var userName: String = ""
    @Published var birthDate: Date = Date()
    @Published var selectedGoals: [String] = []
    @Published var relationshipDuration: AppUser.RelationshipDuration = .oneToThreeYears
    @Published var relationshipImprovement: String = ""
    @Published var selectedImprovements: [String] = []
    @Published var questionMode: String = "🔄 Questions variées"
    @Published var isLoading: Bool = false
    @Published var shouldSkipSubscription: Bool = false
    @Published var shouldShowPartnerConnectionSuccess = false
    @Published var connectedPartnerName: String = ""
    @Published var relationshipStartDate: Date?
    @Published var profileImage: UIImage?
    @Published var currentLocation: UserLocation?
    
    var appState: AppState?
    private var cancellables = Set<AnyCancellable>()
    private var isCompletingSubscription = false
    
    // Options pour les objectifs de relation
    var relationshipGoals: [String] {
        [
            "goal_create_connection".localized,
            "goal_talk_avoided_subjects".localized,
            "goal_increase_passion".localized,
            "goal_share_more_laughs".localized,
            "goal_find_complicity".localized
        ]
    }

    // Options pour l'amélioration de la relation
    var relationshipImprovements: [String] {
        [
            "improvement_create_strong_moment".localized,
            "improvement_revive_connection".localized,
            "improvement_break_routine".localized,
            "improvement_say_unsaid".localized
        ]
    }
    
    // Options pour le mode de questions
    let questionModes = [
        "🎯 Sérieux",
        "🎉 Fun",
        "🌶️ Hot et Sensuel",
        "💫 Profond"
    ]
    
    // Propriété calculée pour la barre de progression
    var progressValue: Double {
        let totalSteps = Double(OnboardingStep.allCases.count)
        let currentIndex = Double(OnboardingStep.allCases.firstIndex(of: currentStep) ?? 0)
        return (currentIndex + 1) / totalSteps
    }
    
    init(appState: AppState) {
        print("🔥 OnboardingViewModel: Initialisation avec AppState")
        self.appState = appState
        
        // Observer les changements d'abonnement
        NotificationCenter.default.addObserver(
            forName: .subscriptionUpdated,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleSubscriptionUpdate()
        }
        
        // NOUVEAU: Observer les connexions partenaire réussies
        NotificationCenter.default.addObserver(
            forName: .partnerConnectionSuccess,
            object: nil,
            queue: .main
        ) { [weak self] notification in
            if let userInfo = notification.userInfo,
               let partnerName = userInfo["partnerName"] as? String {
                self?.showPartnerConnectionSuccess(partnerName: partnerName)
            }
        }
    }
    
    init() {
        print("🔥 OnboardingViewModel: Initialisation sans AppState")
        self.appState = nil
    }
    
    func updateAppState(_ newAppState: AppState) {
        print("🔥 OnboardingViewModel: Mise à jour AppState")
        self.appState = newAppState
    }
    
    func nextStep() {
        let stepNumber = OnboardingStep.allCases.firstIndex(of: currentStep) ?? 0
        
        // 📊 Analytics: Progression onboarding
        Analytics.logEvent("onboarding_etape", parameters: ["etape": stepNumber])
        print("📊 Événement Firebase: onboarding_etape - étape: \(stepNumber)")
        
        switch currentStep {
        case .relationshipGoals:
            currentStep = .relationshipDate
        case .relationshipDate:
            currentStep = .relationshipImprovement
        case .relationshipImprovement:
            currentStep = .authentication
        case .authentication:
            currentStep = .displayName
        case .displayName:
            // Toujours permettre de passer à profilePhoto, même avec nom vide (auto-génération)
            print("🔥 OnboardingViewModel: displayName -> profilePhoto (nom: '\(userName)')")
            currentStep = .profilePhoto
        case .profilePhoto:
            currentStep = .completion
        case .completion:
            currentStep = .loading
            completeDataCollection()
        case .loading:
            if shouldSkipSubscription {
                finalizeOnboarding(withSubscription: true)
            } else {
                currentStep = .partnerCode
            }
        case .partnerCode:
            currentStep = .categoriesPreview
        case .categoriesPreview:
            currentStep = .fitnessIntro
        case .fitnessIntro:
            currentStep = .fitnessIntro2
        case .fitnessIntro2:
            currentStep = .dailyQuestionNotification
        case .dailyQuestionNotification:
            currentStep = .subscription
        case .subscription:
            // L'onboarding doit être finalisé via skipSubscription() ou completeSubscription()
            break
        }
    }
    
    func previousStep() {
        print("🔥 OnboardingViewModel: Retour à l'étape précédente depuis \(currentStep)")
        switch currentStep {
        case .relationshipGoals:
            print("🔥 OnboardingViewModel: Déjà à la première étape")
            break
        case .relationshipDate:
            currentStep = .relationshipGoals
        case .relationshipImprovement:
            currentStep = .relationshipDate
        case .authentication:
            currentStep = .relationshipImprovement
        case .displayName:
            currentStep = .authentication
        case .profilePhoto:
            // Retour conditionnel : si on a un nom Apple, retourner à auth, sinon à displayName
            if let appleDisplayName = AuthenticationService.shared.appleUserDisplayName, !appleDisplayName.isEmpty {
                currentStep = .authentication
            } else {
                currentStep = .displayName
            }
        case .completion:
            currentStep = .profilePhoto
        case .loading:
            currentStep = .completion
        case .partnerCode:
            currentStep = .loading
        case .categoriesPreview:
            currentStep = .partnerCode
        case .fitnessIntro:
            currentStep = .categoriesPreview
        case .fitnessIntro2:
            currentStep = .fitnessIntro
        case .dailyQuestionNotification:
            currentStep = .fitnessIntro2
        case .subscription:
            print("🔥 OnboardingViewModel: Impossible de revenir en arrière depuis l'abonnement")
            break
        }
        print("🔥 OnboardingViewModel: Nouvelle étape: \(currentStep)")
    }
    
    func toggleGoal(_ goal: String) {
        if selectedGoals.contains(goal) {
            selectedGoals.removeAll { $0 == goal }
        } else {
            selectedGoals.append(goal)
        }
    }
    
    private func completeDataCollection() {
        print("🔥 OnboardingViewModel: Début de la finalisation de la collecte de données")
        isLoading = true
        
        // Simuler un traitement
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            print("🔥 OnboardingViewModel: Fin du chargement simulé")
            self.isLoading = false
            self.currentStep = .partnerCode
        }
    }
    
    func completeAuthentication() {
        print("🔥 OnboardingViewModel: Authentification terminée")
        print("🔥 OnboardingViewModel: Attente du traitement Apple Name...")
        
        // NOUVEAU: Attendre que AuthenticationService ait fini de traiter appleUserDisplayName
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            self.checkAppleNameAndProceed()
        }
    }
    
    private func checkAppleNameAndProceed() {
        print("🔥 OnboardingViewModel: Vérification nom Apple...")
        
        let appleDisplayName = AuthenticationService.shared.appleUserDisplayName
        print("🔥 OnboardingViewModel: appleUserDisplayName = '\(appleDisplayName ?? "nil")'")
        print("🔥 OnboardingViewModel: isEmpty = \(appleDisplayName?.isEmpty ?? true)")
        
        // Vérifier si Apple a fourni un nom d'affichage
        if let appleDisplayName = appleDisplayName, !appleDisplayName.isEmpty {
            print("🔥 OnboardingViewModel: ✅ Nom Apple fourni (\(appleDisplayName)) - Skip DisplayNameStepView")
            // Utiliser le nom fourni par Apple
            self.userName = appleDisplayName
            print("🔥 OnboardingViewModel: userName défini à: '\(self.userName)'")
            // Passer directement à la photo de profil
            print("🔥 OnboardingViewModel: Navigation vers .profilePhoto")
            self.currentStep = .profilePhoto
        } else {
            print("🔥 OnboardingViewModel: ❌ Aucun nom Apple fourni - Afficher DisplayNameStepView")
            // Pas de nom fourni, demander un pseudonyme
            print("🔥 OnboardingViewModel: Navigation vers .displayName")
            self.currentStep = .displayName
        }
    }
    
    func skipSubscription() {
        print("🔥 OnboardingViewModel: Abonnement ignoré, finalisation")
        print("⏭️ Onboarding sans abonnement")
        finalizeOnboarding(withSubscription: false)
    }
    
    func completeSubscription() {
        // Protection contre les doubles appels
        guard !isCompletingSubscription else {
            print("🔥 OnboardingViewModel: ⚠️ Appel ignoré - Finalisation déjà en cours")
            return
        }
        
        isCompletingSubscription = true
        print("🔥 OnboardingViewModel: Abonnement terminé, finalisation")
        print("✅ Onboarding avec abonnement")
        finalizeOnboarding(withSubscription: true)
    }
    
    func finalizeOnboarding(withSubscription isSubscribed: Bool = false) {
        print("🔥 OnboardingViewModel: Finalisation complète de l'onboarding")
        print("🎯 Finalisation onboarding")
        
        // NOUVEAU: Désactiver l'overlay de connexion partenaire de l'onboarding
        // car MainView va prendre le relais
        if shouldShowPartnerConnectionSuccess {
            print("🔥 OnboardingViewModel: Désactivation overlay connexion partenaire (MainView prend le relais)")
            shouldShowPartnerConnectionSuccess = false
        }
        
        #if DEBUG
        print("🔥 OnboardingViewModel: Création de l'utilisateur avec:")
        print("  - Nom: \(userName)")
        print("  - Objectifs: \(selectedGoals)")
        print("  - Durée de relation: \(relationshipDuration)")
        print("  - Amélioration souhaitée: \(selectedImprovements)")
        print("  - Mode de questions: \(questionMode)")
        print("  - Abonné: \(isSubscribed)")
        #else
        print("🔥 OnboardingViewModel: Création de l'utilisateur terminée")
        #endif
        
        print("👤 Création utilisateur final")
        
        // Convertir le tableau d'améliorations en string pour Firebase
        let improvementString = selectedImprovements.joined(separator: ", ")
        
        // 🔧 CORRECTION: Préserver les données de connexion partenaire existantes
        FirebaseService.shared.finalizeOnboardingWithPartnerData(
            name: userName,
            relationshipGoals: selectedGoals,
            relationshipDuration: relationshipDuration,
            relationshipImprovement: improvementString.isEmpty ? nil : improvementString,
            questionMode: questionMode.isEmpty ? nil : questionMode,
            isSubscribed: isSubscribed,
            relationshipStartDate: relationshipStartDate,
            profileImage: profileImage,
            currentLocation: currentLocation
        ) { [weak self] success, user in
            DispatchQueue.main.async {
                guard let self = self else { return }
                
                if success, let user = user {
                    print("✅ OnboardingViewModel: Finalisation réussie avec préservation données partenaire")
                            print("✅ Utilisateur créé avec succès")
                    
                    guard let appState = self.appState else {
                        print("❌ OnboardingViewModel: AppState manquant!")
                        NSLog("❌❌❌ ONBOARDING: APPSTATE MANQUANT!")
                        return
                    }
                    
                    print("🔥 OnboardingViewModel: Mise à jour de l'utilisateur via AppState")
                            print("📱 Mise à jour interface")
                    
                    // 📊 Analytics: Onboarding terminé
                    Analytics.logEvent("onboarding_complete", parameters: [:])
                    print("📊 Événement Firebase: onboarding_complete")
                    
                    // NOUVEAU: Marquer la fin du processus d'onboarding dans Firebase et AppState
                    FirebaseService.shared.completeOnboardingProcess()
                    appState.isOnboardingInProgress = false
                    // Réinitialisation des flags d'onboarding
                    
                    appState.updateUser(user)
                    appState.completeOnboarding()
                    print("🔥 OnboardingViewModel: Onboarding terminé")
                            print("🎉 Onboarding terminé avec succès")
                } else {
                    print("❌ OnboardingViewModel: Erreur lors de la finalisation")
                    NSLog("❌❌❌ ONBOARDING: ERREUR FINALISATION!")
                }
                
                // Reset du flag de protection
                self.isCompletingSubscription = false
            }
        }
    }
    
    // DEPRECATED: Ces méthodes ne sont plus utilisées avec le nouveau flux
    func completeOnboardingAfterSubscription() {
        print("⚠️ OnboardingViewModel: DEPRECATED - completeOnboardingAfterSubscription appelée")
        completeAuthentication()
    }
    
    // NOUVEAU: Méthode pour passer l'abonnement suite à un héritage
    func skipSubscriptionDueToInheritance() {
        print("🔥 OnboardingViewModel: Abonnement hérité du partenaire premium - skip subscription")
        shouldSkipSubscription = true
    }
    
    // NOUVEAU: Méthode pour réinitialiser le flag de skip si nécessaire
    func resetSubscriptionSkip() {
        print("🔥 OnboardingViewModel: Reset flag skip subscription")
        shouldSkipSubscription = false
    }
    
    func handleSubscriptionUpdate() {
        print("🔥 OnboardingViewModel: Mise à jour d'abonnement détectée")
        // Gérer les mises à jour d'abonnement si nécessaire
    }
    
    func showPartnerConnectionSuccess(partnerName: String) {
        print("🎉 OnboardingViewModel: Affichage message connexion pour: \(partnerName)")
        connectedPartnerName = partnerName
        shouldShowPartnerConnectionSuccess = true
    }
    
    func dismissPartnerConnectionSuccess() {
        print("🎉 OnboardingViewModel: Fermeture message connexion")
        shouldShowPartnerConnectionSuccess = false
        connectedPartnerName = ""
    }
}

File: ViewModels/OnboardingViewModel.kt
package com.love2loveapp.core.viewmodels.onboarding

import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.ktx.Firebase
import com.love2loveapp.core.common.AppException
import com.love2loveapp.core.common.Result
import com.love2loveapp.model.AppConstants
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.Date

/**
 * Placeholders de ressources:
 * strings.xml doit contenir (exemples)
 *
 * <string name="goal_create_connection">Créer une connexion</string>
 * <string name="goal_talk_avoided_subjects">Parler des sujets évités</string>
 * <string name="goal_increase_passion">Augmenter la passion</string>
 * <string name="goal_share_more_laughs">Partager plus de rires</string>
 * <string name="goal_find_complicity">Trouver la complicité</string>
 *
 * <string name="improvement_create_strong_moment">Créer un moment fort</string>
 * <string name="improvement_revive_connection">Raviver la connexion</string>
 * <string name="improvement_break_routine">Casser la routine</string>
 * <string name="improvement_say_unsaid">Dire l’inavoué</string>
 */

// --- Événements globaux (remplacement simple de NotificationCenter) ---
object SubscriptionEvents {
    val updates = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
}
object PartnerEvents {
    // émet le nom du partenaire connecté
    val connectionSuccess = MutableSharedFlow<String>(extraBufferCapacity = 1)
}

// --- Types domaine “placeholder” à mapper sur tes vrais types ---
data class UserLocation(
    val latitude: Double,
    val longitude: Double,
    val address: String? = null,
    val city: String? = null,
    val country: String? = null,
    val lastUpdated: Long? = null
)

enum class RelationshipDuration {
    LessThanSixMonths, SixMonthsToOneYear, OneToThreeYears, ThreeToFiveYears, MoreThanFiveYears
}

data class AppUser(
    val id: String,
    val name: String,
    val isSubscribed: Boolean
)

interface AppState {
    var isOnboardingInProgress: Boolean
    fun updateUser(user: AppUser)
    fun completeOnboarding()
}

interface FirebaseService {
    /**
     * Doit créer/mettre à jour l’utilisateur en préservant les données partenaire si présentes.
     * Retourne (success, user)
     */
    suspend fun finalizeOnboardingWithPartnerData(
        name: String,
        relationshipGoals: List<String>,
        relationshipDuration: RelationshipDuration,
        relationshipImprovement: String?,
        questionMode: String?,
        isSubscribed: Boolean,
        relationshipStartDate: Date?,
        profileImage: Bitmap?,
        currentLocation: UserLocation?
    ): Pair<Boolean, AppUser?>

    fun completeOnboardingProcess()
}

// --- ViewModel ---
class OnboardingViewModel(
    private val firebaseService: FirebaseService,
    var appState: AppState? = null
) : ViewModel() {

    // Étapes (ordre identique à Swift)
    enum class OnboardingStep {
        relationshipGoals,
        relationshipImprovement,
        relationshipDate,
        communicationEvaluation,
        discoveryTime,
        listening,
        confidence,
        complicity,
        authentication,
        displayName,
        profilePhoto,
        completion,
        loading,
        partnerCode,
        questionsIntro,
        categoriesPreview,
        subscription
    }

    // --- State (équivalents @Published) ---
    val currentStep = MutableStateFlow(OnboardingStep.relationshipGoals)
    val userName = MutableStateFlow("")
    val birthDate = MutableStateFlow(Date())
    val selectedGoals = MutableStateFlow<List<String>>(emptyList())
    val relationshipDuration = MutableStateFlow(RelationshipDuration.OneToThreeYears)
    val relationshipImprovement = MutableStateFlow("")
    val selectedImprovements = MutableStateFlow<List<String>>(emptyList())
    val questionMode = MutableStateFlow("🔄 Questions variées")
    val isLoading = MutableStateFlow(false)
    val shouldSkipSubscription = MutableStateFlow(false)
    val shouldShowPartnerConnectionSuccess = MutableStateFlow(false)
    val connectedPartnerName = MutableStateFlow("")
    val relationshipStartDate = MutableStateFlow<Date?>(null)
    val profileImage = MutableStateFlow<Bitmap?>(null)
    val currentLocation = MutableStateFlow<UserLocation?>(null)

    // Évaluation du couple
    val communicationRating = MutableStateFlow("")
    val discoveryTimeAnswer = MutableStateFlow("")
    val listeningAnswer = MutableStateFlow("")
    val confidenceAnswer = MutableStateFlow("")
    val complicityAnswer = MutableStateFlow("")

    private val analytics = Firebase.analytics
    private var isCompletingSubscriptionGuard = false

    // --- Localisation via strings.xml : listes de @StringRes puis helpers de résolution ---
    private val relationshipGoalsRes: List<Int> = listOf(
        R.string.goal_create_connection,
        R.string.goal_talk_avoided_subjects,
        R.string.goal_increase_passion,
        R.string.goal_share_more_laughs,
        R.string.goal_find_complicity
    )

    private val relationshipImprovementsRes: List<Int> = listOf(
        R.string.improvement_create_strong_moment,
        R.string.improvement_revive_connection,
        R.string.improvement_break_routine,
        R.string.improvement_say_unsaid
    )

    /** Récupère la liste localisée des objectifs depuis un Context */
    fun relationshipGoals(context: Context): List<String> =
        relationshipGoalsRes.map(context::getString)

    /** Récupère la liste localisée des améliorations depuis un Context */
    fun relationshipImprovements(context: Context): List<String> =
        relationshipImprovementsRes.map(context::getString)

    // Modes de questions (texte direct ici ; tu peux aussi passer par strings.xml si tu préfères)
    val questionModes: List<String> = listOf(
        "🎯 Sérieux",
        "🎉 Fun",
        "🌶️ Hot et Sensuel",
        "💫 Profond"
    )

    /** Progression 0..1 (réactive) */
    val progress: StateFlow<Float> = currentStep
        .map { step ->
            val total = OnboardingStep.values().size.toFloat()
            val index = OnboardingStep.values().indexOf(step).toFloat()
            (index + 1f) / total
        }
        .stateIn(viewModelScope, SharingStarted.Eagerly, 1f / OnboardingStep.values().size)

    init {
        Log.d("OnboardingVM", "🔥 Initialisation ViewModel")
        // Observateurs équivalents à NotificationCenter
        SubscriptionEvents.updates
            .onEach { handleSubscriptionUpdate() }
            .launchIn(viewModelScope)

        PartnerEvents.connectionSuccess
            .onEach { showPartnerConnectionSuccess(it) }
            .launchIn(viewModelScope)
    }

    fun updateAppState(newAppState: AppState) {
        Log.d("OnboardingVM", "🔥 Mise à jour AppState")
        this.appState = newAppState
    }

    fun nextStep() {
        val stepIndex = OnboardingStep.values().indexOf(currentStep.value)
        analytics.logEvent("onboarding_etape") { param("etape", stepIndex.toLong()) }
        Log.d("OnboardingVM", "📊 Firebase event: onboarding_etape - étape: $stepIndex")

        when (currentStep.value) {
            OnboardingStep.relationshipGoals -> currentStep.value = OnboardingStep.relationshipImprovement
            OnboardingStep.relationshipImprovement -> currentStep.value = OnboardingStep.relationshipDate
            OnboardingStep.relationshipDate -> currentStep.value = OnboardingStep.communicationEvaluation
            OnboardingStep.communicationEvaluation -> currentStep.value = OnboardingStep.discoveryTime
            OnboardingStep.discoveryTime -> currentStep.value = OnboardingStep.listening
            OnboardingStep.listening -> currentStep.value = OnboardingStep.confidence
            OnboardingStep.confidence -> currentStep.value = OnboardingStep.complicity
            OnboardingStep.complicity -> currentStep.value = OnboardingStep.authentication
            OnboardingStep.authentication -> currentStep.value = OnboardingStep.displayName
            OnboardingStep.displayName -> {
                Log.d("OnboardingVM", "🔥 displayName -> profilePhoto (nom: '${userName.value}')")
                currentStep.value = OnboardingStep.profilePhoto
            }
            OnboardingStep.profilePhoto -> currentStep.value = OnboardingStep.completion
            OnboardingStep.completion -> {
                currentStep.value = OnboardingStep.loading
                completeDataCollection()
            }
            OnboardingStep.loading -> {
                if (shouldSkipSubscription.value) {
                    finalizeOnboarding(withSubscription = true)
                } else {
                    currentStep.value = OnboardingStep.partnerCode
                }
            }
            OnboardingStep.partnerCode -> currentStep.value = OnboardingStep.questionsIntro
            OnboardingStep.questionsIntro -> currentStep.value = OnboardingStep.categoriesPreview
            OnboardingStep.categoriesPreview -> currentStep.value = OnboardingStep.subscription
            OnboardingStep.subscription -> {
                // finalisation via skipSubscription() ou completeSubscription()
            }
        }
    }

    fun previousStep() {
        Log.d("OnboardingVM", "🔥 Retour depuis ${currentStep.value}")
        currentStep.value = when (currentStep.value) {
            OnboardingStep.relationshipGoals -> OnboardingStep.relationshipGoals
            OnboardingStep.relationshipImprovement -> OnboardingStep.relationshipGoals
            OnboardingStep.relationshipDate -> OnboardingStep.relationshipImprovement
            OnboardingStep.communicationEvaluation -> OnboardingStep.relationshipDate
            OnboardingStep.discoveryTime -> OnboardingStep.communicationEvaluation
            OnboardingStep.listening -> OnboardingStep.discoveryTime
            OnboardingStep.confidence -> OnboardingStep.listening
            OnboardingStep.complicity -> OnboardingStep.confidence
            OnboardingStep.authentication -> OnboardingStep.complicity
            OnboardingStep.displayName -> OnboardingStep.authentication
            OnboardingStep.profilePhoto -> {
                // Google Only : si un displayName Google existe on peut estimer revenir à auth, sinon displayName
                val googleName = FirebaseAuth.getInstance().currentUser?.displayName
                if (!googleName.isNullOrBlank()) OnboardingStep.authentication else OnboardingStep.displayName
            }
            OnboardingStep.completion -> OnboardingStep.profilePhoto
            OnboardingStep.loading -> OnboardingStep.completion
            OnboardingStep.partnerCode -> OnboardingStep.loading
            OnboardingStep.questionsIntro -> OnboardingStep.partnerCode
            OnboardingStep.categoriesPreview -> OnboardingStep.questionsIntro
            OnboardingStep.subscription -> {
                Log.d("OnboardingVM", "🔥 Impossible de revenir depuis subscription")
                OnboardingStep.subscription
            }
        }
        Log.d("OnboardingVM", "🔥 Nouvelle étape: ${currentStep.value}")
    }

    fun toggleGoal(goal: String) {
        val current = selectedGoals.value.toMutableList()
        if (current.contains(goal)) current.remove(goal) else current.add(goal)
        selectedGoals.value = current
    }

    private fun completeDataCollection() {
        Log.d("OnboardingVM", "🔥 Début finalisation collecte")
        viewModelScope.launch {
            isLoading.value = true
            delay(AppConstants.Onboarding.DATA_COLLECTION_DELAY_MS)
            isLoading.value = false
            currentStep.value = OnboardingStep.partnerCode
            Log.d("OnboardingVM", "🔥 Fin du chargement simulé -> partnerCode")
        }
    }

    /** Google Sign-In uniquement */
    fun completeAuthentication() {
        Log.d("OnboardingVM", "🔥 Authentification terminée (Google)")
        viewModelScope.launch {
            // Lecture du displayName Google si dispo
            checkGoogleNameAndProceed()
        }
    }

    private fun checkGoogleNameAndProceed() {
        val googleDisplayName = FirebaseAuth.getInstance().currentUser?.displayName
        Log.d("OnboardingVM", "🔥 googleUserDisplayName='${googleDisplayName ?: "null"}'")

        if (!googleDisplayName.isNullOrBlank()) {
            Log.d("OnboardingVM", "✅ Nom Google fourni ($googleDisplayName) - Skip DisplayName step")
            userName.value = googleDisplayName
            currentStep.value = OnboardingStep.profilePhoto
        } else {
            Log.d("OnboardingVM", "❌ Aucun nom Google - Aller vers DisplayName")
            currentStep.value = OnboardingStep.displayName
        }
    }

    fun skipSubscription() {
        Log.d("OnboardingVM", "🔥 Abonnement ignoré -> finalisation")
        finalizeOnboarding(withSubscription = false)
    }

    fun completeSubscription() {
        if (isCompletingSubscriptionGuard) {
            Log.d("OnboardingVM", "🔥 Appel ignoré, finalisation déjà en cours")
            return
        }
        isCompletingSubscriptionGuard = true
        Log.d("OnboardingVM", "🔥 Abonnement terminé -> finalisation")
        finalizeOnboarding(withSubscription = true)
    }
    
    // === Result<T> based operations ===
    
    /**
     * Valide les données d'onboarding avec Result<T>
     */
    fun validateOnboardingData(): Result<Boolean> = try {
        when {
            userName.value.isBlank() -> Result.error(AppException.Data.ValidationError("user_name"))
            selectedGoals.value.isEmpty() -> Result.error(AppException.Data.ValidationError("relationship_goals"))
            relationshipStartDate.value == null -> Result.error(AppException.Data.ValidationError("relationship_date"))
            else -> Result.success(true)
        }
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }
    
    /**
     * État de progression avec Result<T>
     */
    fun getProgressResult(): Result<Float> = try {
        val progressValue = progress.value
        if (progressValue in 0f..1f) {
            Result.success(progressValue)
        } else {
            Result.error(AppException.Generic("Invalid progress value: $progressValue"))
        }
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }
    
    /**
     * Finalisation avec Result<T>
     */
    suspend fun finalizeOnboardingResult(withSubscription: Boolean): Result<Unit> = try {
        val validation = validateOnboardingData()
        if (validation is Result.Error) {
            return validation
        }
        
        val improvementString = selectedImprovements.value.joinToString(", ").ifBlank { null }
        
        val (success, user) = firebaseService.finalizeOnboardingWithPartnerData(
            name = userName.value,
            relationshipGoals = selectedGoals.value,
            relationshipDuration = relationshipDuration.value,
            relationshipImprovement = improvementString,
            questionMode = questionMode.value.ifBlank { null },
            isSubscribed = withSubscription,
            relationshipStartDate = relationshipStartDate.value,
            profileImage = profileImage.value,
            currentLocation = currentLocation.value
        )
        
        if (success && user != null) {
            firebaseService.completeOnboardingProcess()
            appState?.let { state ->
                state.isOnboardingInProgress = false
                state.updateUser(user)
                state.completeOnboarding()
            }
            Result.success(Unit)
        } else {
            Result.error(AppException.Generic("Onboarding finalization failed"))
        }
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }

    fun skipSubscriptionDueToInheritance() {
        Log.d("OnboardingVM", "🔥 Abonnement hérité du partenaire -> skip subscription")
        shouldSkipSubscription.value = true
    }

    fun resetSubscriptionSkip() {
        Log.d("OnboardingVM", "🔥 Reset skip subscription")
        shouldSkipSubscription.value = false
    }

    fun handleSubscriptionUpdate() {
        Log.d("OnboardingVM", "🔥 Mise à jour d’abonnement détectée")
        // À implémenter si besoin (sync UI/serveur)
    }

    fun showPartnerConnectionSuccess(partnerName: String) {
        Log.d("OnboardingVM", "🎉 Connexion partenaire: $partnerName")
        connectedPartnerName.value = partnerName
        shouldShowPartnerConnectionSuccess.value = true
    }

    fun dismissPartnerConnectionSuccess() {
        Log.d("OnboardingVM", "🎉 Fermeture message connexion")
        shouldShowPartnerConnectionSuccess.value = false
        connectedPartnerName.value = ""
    }

    private fun finalizeOnboarding(withSubscription: Boolean) {
        Log.d("OnboardingVM", "🔥 Finalisation complète de l’onboarding")
        // Désactiver l’overlay de connexion partenaire : l’écran principal prendra le relais
        if (shouldShowPartnerConnectionSuccess.value) {
            Log.d("OnboardingVM", "🔥 Désactivation overlay partenaire (MainView prend le relais)")
            shouldShowPartnerConnectionSuccess.value = false
        }

        analytics.logEvent("onboarding_complete") {}
        Log.d("OnboardingVM", "📊 Firebase event: onboarding_complete")

        viewModelScope.launch {
            val improvementString = selectedImprovements.value.joinToString(", ").ifBlank { null }

            val (success, user) = firebaseService.finalizeOnboardingWithPartnerData(
                name = userName.value,
                relationshipGoals = selectedGoals.value,
                relationshipDuration = relationshipDuration.value,
                relationshipImprovement = improvementString,
                questionMode = questionMode.value.ifBlank { null },
                isSubscribed = withSubscription,
                relationshipStartDate = relationshipStartDate.value,
                profileImage = profileImage.value,
                currentLocation = currentLocation.value
            )

            if (success && user != null) {
                Log.d("OnboardingVM", "✅ Utilisateur créé avec succès (préservation partenaire OK)")
                firebaseService.completeOnboardingProcess()

                appState?.let { state ->
                    state.isOnboardingInProgress = false
                    state.updateUser(user)
                    state.completeOnboarding()
                } ?: run {
                    Log.e("OnboardingVM", "❌ AppState manquant")
                }
            } else {
                Log.e("OnboardingVM", "❌ Erreur lors de la finalisation")
            }

            isCompletingSubscriptionGuard = false
        }
    }
}
